src/main/java/io/quarkus/security/AuthenticationFailedException.java

```diff
@@ -11,19 +13,60 @@
  * due to bad credentials vs some other form of internal failure.
  */
 public final class AuthenticationFailedException extends SecurityException implements AuthenticationException {
+
+    private final Map<String, Object> attributes;
+
     public AuthenticationFailedException() {
+        this(null, null, null);
+    }
 
+    public AuthenticationFailedException(Map<String, Object> attributes) {
+        this(null, null, attributes);
     }
 
     public AuthenticationFailedException(String errorMessage) {
-        this(errorMessage, null);
+        this(errorMessage, null, null);
+    }
+
+    public AuthenticationFailedException(String errorMessage, Map<String, Object> attributes) {
+        this(errorMessage, null, attributes);
     }
 
     public AuthenticationFailedException(Throwable cause) {
-        this(null, cause);
+        this(null, cause, null);
+    }
+
+    public AuthenticationFailedException(Throwable cause, Map<String, Object> attributes) {
+        this(null, cause, attributes);
     }
 
     public AuthenticationFailedException(String errorMessage, Throwable cause) {
+        this(errorMessage, cause, null);
+    }
+
+    public AuthenticationFailedException(String errorMessage, Throwable cause, Map<String, Object> attributes) {
         super(errorMessage, cause);
+        this.attributes = attributes == null ? null : Map.copyOf(attributes);
+    }
+
+    /**
+     * Provides an attribute that allows to better understand the cause of the authentication failure.
+     *
+     * @param name The authentication failure attribute name
+     * @return The authentication attribute value or null if the attribute with this name does not exist
+     * @param <T> The type of the authentication failure attribute
+     */
+    @SuppressWarnings("unchecked")
+    public <T> T getAttribute(String name) {
+        return attributes != null ? ((T) attributes.get(name)) : null;
+    }
+
+    /**
+     * Provides attributes that allows to better understand the cause of the authentication failure.
+     *
+     * @return authentication failure attributes; never null
+     */
+    public Map<String, Object> getAttributes() {
+        return attributes == null ? Map.of() : attributes;
```

Comment:
I believe `attributes` should be returned as unmodifiable

Reply:
`Map.copy` returns immutable map, therefore they are unmodifiable as well.

Reply:
correction `Map.copy` returns unmodifiable, not sure why I thought it returns immutable. I think I read I assumed it creates a new map and doesn't rely on the underlying one (so it is smaller and can't be changed), but javadoc doesn't say it.

Reply:
and `Map.copy` is used in the constructor (that is what I am referring to).

Reply:
@michalvavrik Sorry, missed it

Reply:
np!